### PR TITLE
feat: onboard new destination dynamic yield

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

Onboard new destination dynamic yield

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/dynamic_yield/transform.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  const payload = constructPayload(message, mappingConfig[ConfigCategory.IDENTIFY.name]);
+  payload.events = [];
+  const eventsObj = {
+    name: 'Identify User',
+    properties: {
+      dyType: 'identify-v1',
+      hashedEmail:
+        message.traits?.email || message.context?.traits?.email
+          ? sha256(message.traits?.email || message.context?.traits?.email)
+          : null,
```
Comment (ItsSudip) [inline src/v0/destinations/dynamic_yield/transform.js:45]: We have to check from the config i.e. if the customer sending hashed data or not.

Comment (aashishmalik) [reply src/v0/destinations/dynamic_yield/transform.js:45]: we can also do checking in code itself, if the provided email is already hashed or not. WDYT?

File: src/v0/destinations/dynamic_yield/util.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    dyType: `${ecomEventsMapping[event].toLowerCase().trim().replace(/\s+/g, '-')}-v1`,
+    value: Number(parseFloat(messageProperties.price).toFixed(2)),
+    currency: messageProperties.currency,
+  };
+  if (event === 'order_completed') {
+    properties.uniqeTransactionId = messageProperties.order_id;
+    properties.value = messageProperties.revenue;
+  } else {
+    properties.quantity = messageProperties.quantity;
+    properties.productId = messageProperties.sku;
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/util.js]: Updated

File: src/v0/destinations/dynamic_yield/util.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+const populateProperties = (messageProperties, event) => {
+  const properties = {
+    dyType: `${ecomEventsMapping[event].toLowerCase().trim().replace(/\s+/g, '-')}-v1`,
+    value: Number(parseFloat(messageProperties.price).toFixed(2)),
+    currency: messageProperties.currency,
+  };
+  if (event === 'order_completed') {
+    properties.uniqeTransactionId = messageProperties.order_id;
+    properties.value = messageProperties.revenue;
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/util.js]: Updated

File: src/v0/destinations/dynamic_yield/transform.js (hunk 1 above)
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/transform.js]: Added hashEmail field in config

File: src/v0/destinations/dynamic_yield/util.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  } else {
+    properties.quantity = messageProperties.quantity;
+    properties.productId = messageProperties.sku || messageProperties.product_id;
+  }
+  if (messageProperties.products && Array.isArray(messageProperties.products)) {
+    properties.cart = [];
+    const { products } = messageProperties;
+    products.forEach((product) => {
+      const cartObj = {};
+      cartObj.productId = product.sku;
```
Comment (ItsSudip) [inline src/v0/destinations/dynamic_yield/util.js:30]:
```suggestion
      cartObj.productId = product.sku || product.product_id;
```

**Review by ItsSudip — APPROVED** [review]

File: src/v0/destinations/dynamic_yield/util.js (hunk 4 above)
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/util.js]: Updated

**Review by ItsSudip — DISMISSED** [review]

**Review by aashishmalik — DISMISSED** [review]

File: src/v0/destinations/dynamic_yield/transform.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    response.body.JSON = removeUndefinedAndNullAndEmptyValues(payload);
+    return response;
+  }
+  throw new TransformationError('Payload could not be populated due to wrong input');
+};
+
+const identifyResponseBuilder = (message, { Config }) => {
+  const { apiKey, hashEmail } = Config;
+
+  const payload = constructPayload(message, mappingConfig[ConfigCategory.IDENTIFY.name]);
```
Comment (krishna2020) [inline src/v0/destinations/dynamic_yield/transform.js:36]: Move the below code L37-L56 to `util.js` under `prepareIdentifyPayload` and add unit test to it

File: src/v0/destinations/dynamic_yield/util.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        itemPrice: product.price,
+      };
+      properties.cart.push(cartObj);
+    });
+  }
+  return properties;
+};
+
+const preparePayload = (message, event) => {
+  const payload = constructPayload(message, mappingConfig[ConfigCategory.TRACK.name]);
```
Comment (krishna2020) [inline src/v0/destinations/dynamic_yield/util.js:41]: Move `constructPayload` to `trackResponseBuilder` & clean the corresponding import of config.js

File: src/v0/destinations/dynamic_yield/util.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+const { ConfigCategory, mappingConfig } = require('./config');
+
+const ecomEventsMapping = {
+  product_added: 'Add to Cart',
+  product_removed: 'Remove from Cart',
+  product_added_to_wishlist: 'Add to Wishlist',
+  order_completed: 'Purchase',
+};
+
+const populateProperties = (messageProperties, event) => {
```
Comment (krishna2020) [inline src/v0/destinations/dynamic_yield/util.js:11]: write unit test for this function

File: test/integrations/destinations/dynamic_yield/processor/data.js
Comment (krishna2020) [inline test/integrations/destinations/dynamic_yield/processor/data.js:1]: Any use-case we have to support `processorTransform` as this is new destination ?

**Comment by krishna2020** [comment]
Support this destination on CDK v2 too & this destination should go live directly on CDK v2 - See whether `dynamic_yield/util.js` can be re-used

File: src/v0/destinations/dynamic_yield/transform.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+const process = async (event) => processEvent(event.message, event.destination);
+
+const processRouterDest = async (inputs, reqMetadata) => {
+  const respList = await simpleProcessRouterDest(inputs, process, reqMetadata);
+  return respList;
+};
+
+module.exports = { process, processRouterDest };
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/transform.js:110]: This is a simple destination (`processTransform` only). We are not enabling `routerTransform` for now. Just have added the code, if we want to enable in future.

File: src/v0/destinations/dynamic_yield/transform.js (hunk 5 above)
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/transform.js]: Done

File: src/v0/destinations/dynamic_yield/util.js (hunk 6 above)
Comment (ujjwal-ab) [reply src/v0/destinations/dynamic_yield/util.js]: Done

**Comment by sanpj2292** [comment]
Since we're going ahead with CDKV2 implementation instead of the native implementation, closing this PR

**Pull request closed without merging**